torch/csrc/autograd/variable.h

```diff
@@ -23,37 +23,34 @@ struct Variable : std::enable_shared_from_this<Variable> {
 
     SavedVariable(const Variable& variable, Function* saved_for)
       : data(variable.data->clone_shallow())
+      , has_grad_fn(variable.grad_fn != nullptr)
       , grad_accumulator(variable.grad_accumulator)
       , version(variable.version_counter->new_saved_ref())
       , requires_grad(variable.requires_grad)
       , is_volatile(false)
       , expected_version(**variable.version_counter) {
-        if (variable.grad_fn.get() == saved_for) {
-          weak_grad_fn = variable.grad_fn;
-        } else {
+        if (variable.grad_fn.get() != saved_for) {
           grad_fn = variable.grad_fn;
         }
       }
 
     std::unique_ptr<thpp::Tensor> data;
-    // The gradient function associated with this node.  If this is
-    // NULL, then this node is a leaf node.
+    // The gradient function associated with this node. If has_grad_fn
+    // is false, then this is a leaf node. Note that the grad_fn is not saved if
+    // it would create a circular reference. In that case, the grad_fn must be
+    // passed in to the unpack function when reconstructing the Variable.
+    bool has_grad_fn;
     std::shared_ptr<Function> grad_fn;
-    // this field is only necessary in case when a grad_fn saves a reference to
-    // one of the outputs of the forward fn. Saving the pointer in grad_fn
-    // would create a reference cycle. If this field is used, grad_fn is
-    // guaranteed to hold a nullptr;
-    std::weak_ptr<Function> weak_grad_fn;
     std::weak_ptr<Function> grad_accumulator;
     std::unique_ptr<VariableVersion> version;
     bool requires_grad;
     bool is_volatile;
     int expected_version;
 
-    std::shared_ptr<Variable> unpack();
+    std::shared_ptr<Variable> unpack(std::shared_ptr<Function> saved_for=nullptr);
```

Comment:
I think it's the easiest thing for now. The other callers are all in C++ implemented autograd functions (like Conv and BatchNorm). None of these currently have to deal with these circular references and it's a little tricky to get a `shared_ptr<Function>` in the backward() context (they only have access to a this).

We shouldn't add `enable_shared_from_this` to the base Function class since it won't work with PyFunction, but if necessary we can add it to non-PyFunction subclasses.